src/bpfprogram.cpp

```diff
@@ -81,17 +85,44 @@ void BpfProgram::set_attach_target(const Probe &probe,
     btf_fun = fun;
   }
 
-  if (btf.get_btf_id(btf_fun, mod, btf_kind) < 0) {
-    const std::string msg = "No BTF found for " + attach_target + ".";
+  std::string err_msg;
+
+  if ((probe.type == ProbeType::fentry || probe.type == ProbeType::fexit) &&
+      mod == "bpf") {
+    auto pair = util::split_string(fun, ':');
```

Comment:
I'm not going to block on this but I don't like that this relies on always doing full expansion of `fentry:bpf` probes, even when the probe is fully specified (like `fentry:bpf:my_prog:1234`). Ideally, what we should do here is introduce a new field on `bpftrace::Probe` and update it from `BPFtrace::generate_probe` and `AttachPoint::create_expansion_copy`. It's some amount of additional code so I understand why we'd want to avoid it. I'm just pointing this out b/c if someone decides to not do expansion for the explicit case in future, the implementation will break.

Reply:
IMHO, the cause of the problem that I'm describing is that we only read the new `AttachPoint::bpf_prog_id` field in `ProbeMatcher`, i.e. we always have to go through it, otherwise the id is not assigned anywhere. And in `ProbeMatcher`, the id is made a part of `func` so that's why we need to do the split on colon here.

Reply:
I see what you mean. Yeah I was bothered a bit by that too but wasn't sure if it was better to add a new field to `Probe` that was so specific to BPF program attachment. The other issue is all the custom code you'd need in `ProbeMatcher` to pass around the found `bpf_prog_id` when you just have a program or sub-program name. TBH, I thought it was kinda nice (except for this part) that you could just make it part of the function name and everything else just worked.

Reply:
@viktormalik My apologies, I got completely confused between the `ast::Probe` and the `struct Probe`. I added a `bpf_prog_id` to `struct Probe` and now the only place we have to do the string split is in `create_expansion_copy`. Let me know if this is what you had in mind.

Reply:
Yes, this is exactly what I had in mind (sorry if I wasn't clear enough). It looks much better now, thanks!